pkg/controller/apiextensions/definer/api.go

```diff
@@ -0,0 +1,109 @@
+/*
+Copyright 2020 The Crossplane Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package definer
+
+import (
+	"context"
+
+	"github.com/pkg/errors"
+	"k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1"
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+	metaapi "k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/crossplane/crossplane-runtime/pkg/meta"
+	"github.com/crossplane/crossplane-runtime/pkg/resource"
+)
+
+const (
+	errDeleteNonControlledCRD = "cannot delete a crd that the definer does not control"
+	errGenerateCRD            = "cannot generate crd for given infrastructure definition"
+)
+
+// NewAPIInfrastructureClient returns a new APIInfrastructureClient.
+func NewAPIInfrastructureClient(client client.Client) Client {
```

Comment:
It requires the `Apply` to be of type patch since there is a generation from the ground up instead of in-place changes. It creates its own `resource.ClientApplicator` to ensure that it's a patch applicator.